src/correction.cc

```diff
@@ -319,17 +319,35 @@ double HashPRNG::evaluate(const std::vector<Variable::Type>& values) const {
 
 Binning::Binning(const JSONObject& json, const Correction& context)
 {
-  std::vector<double> edges;
-  for (const auto& item : json.getRequired<rapidjson::Value::ConstArray>("edges")) {
-    if ( ! item.IsDouble() ) { throw std::runtime_error("Invalid edges array type"); }
-    double val = item.GetDouble();
-    if ( edges.size() > 0 && edges.back() >= val ) { throw std::runtime_error("binning edges are not monotone increasing"); }
-    edges.push_back(val);
-  }
   const auto& content = json.getRequired<rapidjson::Value::ConstArray>("content");
-  if ( edges.size() != content.Size() + 1 ) {
-    throw std::runtime_error("Inconsistency in Binning: number of content nodes does not match binning");
+
+  // set bins_
+  const auto &edgesObj = json.getRequiredValue("edges");
+  if ( edgesObj.IsArray() ) { // non-uniform binning
+    std::vector<double> edges;
+    rapidjson::Value::ConstArray edgesArr = edgesObj.GetArray();
+    for (const auto& edge : edgesArr) {
+      if ( ! edge.IsDouble() ) { throw std::runtime_error("Invalid edges array type"); }
+      double val = edge.GetDouble();
+      if ( edges.size() > 0 && edges.back() >= val ) { throw std::runtime_error("binning edges are not monotone increasing"); }
+      edges.push_back(val);
+    }
+    if ( edges.size() != content.Size() + 1 ) {
+      throw std::runtime_error("Inconsistency in Binning: number of content nodes does not match binning");
+    }
+    _NonUniformBins bins{std::move(edges)};
+    bins_ = std::move(bins);
+  } else { // must be UniformBinning
+    const JSONObject uniformBins{edgesObj.GetObject()};
+    const auto n = uniformBins.getRequired<uint64_t>("n");
```

Comment:
Check for n=0 here (since we don't always run the json through pydantic)

Reply:
done!